src/libsyntax/ext/tt/quoted.rs

```diff
@@ -422,10 +422,35 @@ where
                                 GateIssue::Language,
                                 explain,
                             );
+                        } else {
+                            // We don't allow `?` as a separator
+                            sess.span_diagnostic.span_err(
+                                span,
+                                "`?` is no longer an accepted separator"
+                            );
                         }
                         return (Some(token::Question), op);
                     }
-                    Ok(Ok(op)) => return (Some(token::Question), op),
+                    Ok(Ok(op)) => {
+                        if !features.macro_at_most_once_rep
```

Comment:
I think the feature gate is not *quite* acting like I expect. My expectation is:

- In Edition 2015:
  - If you do `$()?[*+]`, then you get "`?` as separator".
  - Else, if you do `$(foo)?[^*+]`, treat as "optional":
    - But if feature gate is not enabled, error
- In Edition 2018:
  - If you do `$(foo)?`, then treat as "optional":
    - But if feature gate is not enabled, error.

My goal is that, when it comes time to stabilize, we can just "constant propagate" as if the feature gate was always enabled and get the final behavior we want:

- in Rust 2015 we fallback to old behavior 
- in Rust 2018 we use new behavior all the time

(Is this the plan? Do we have an official, FCP'd proposal somewhere, anyway?)

Reply:
> (Is this the plan? Do we have an official, FCP'd proposal somewhere, anyway?)

No, we don't TMK

> in Rust 2015 we fallback to old behavior

Given that this is behavior that will only last for another couple of months, my preference is to
- just keep this unstable on 2015 and only enable in 2018
- make the feature flag always turn on 2018 behavior

Reply:
So to clarify:

- In edition 2015:
    - Without feature flag: `?` is a separator. `?` is _not_ a Kleene op.
    - With feature flag: same as 2018
- In edition 2018
    - `?` is _not_ a separator. `?` is a Kleene op that does not accept a separator.

Reply:
I would prefer this:

- In edition 2015:
  - With *or without* feature flag: `?` is a separator and there is no `?` 
  - However, using `?` as a separator issues a `rust_2018_migration` warning
- In edition 2018:
  - `?` is always a separator
    - and feature-gate is required until we stabilize

Point is, as before, that the feature-gate should just be a "error or no error" filter -- the behavior should model what we eventually want when all is said and done otherwise. And we never want Rust 2015 to support `?`.

Reply:
> In edition 2018:
>    ? is always a separator 

You mean "in 2018, `?` is always a kleene op", right?

Reply:
(I'd sure hope so heh!)